torch/_dynamo/variables/builtin.py

```diff
@@ -162,6 +162,122 @@ def _reversible_binops():
         }
         return fns
 
+    @staticmethod
+    @functools.lru_cache(None)
+    def _binop_handlers():
+        # Multiple dispatch mechanism defining custom binop behavior for certain type
+        # combinations. Handlers are attempted in order, and will be used if the type checks
+        # match. They are expected to have the signature:
+        # fn(tx, arg0: VariableTracker, arg1: VariableTracker, options) -> VariableTracker
+
+        # Override table contains: op_fn -> [list of handlers]
+        op_handlers = {}
+        for (
+            op,
+            (forward_name, reverse_name),
+        ) in BuiltinVariable._reversible_binops().items():
+            handlers = []
+
+            # User-defined args (highest precedence)
+            def user_defined_handler(
+                tx, a, b, options, forward_name=forward_name, reverse_name=reverse_name
+            ):
+                # Manually handle reversing logic if needed (e.g. call __radd__)
+                if isinstance(a, UserDefinedVariable):
+                    return a.call_method(tx, forward_name, [b], {})
+                else:
+                    return b.call_method(tx, reverse_name, [a], {})
```

Comment:
This is technically incorrect because when you have VariableTracker + UserDefinedVariable, whether or not you end up going to `__radd__` depends on the specifics of VariableTracker. For something like int, it will always go to the user defined variable. But here is a naughty example:

```
import torch

class A(int):
    def __radd__(self, other):
        print("woof")

torch.randn(3) + A(3)
```

this prints nothing, because Tensor.__add__ only does a subtype test against int and will ignore the subclass. (Arguably Tensor is wrong in doing it this way, but c'est la vie.)

Reply:
Oh interesting! Nice example, and it sadly invalidates my assumption that this is okay to do. It sounds like we can't get away from actually running the logic and checking for `NotImplemented` -> remove the node from the graph?

Reply:
Well, not quite. In this case, you need to know something about how `Tensor.__add__` is implemented. Now that you know that it never returns NotImplemented if the RHS is an int subtype, you could hard code that here.

Reply:
@ezyang should I do something about this before landing or just drop in a TODO?

Reply:
@ezyang Realized your test case will not hit the logic in this PR, since it's limited to binary op invocation with *non-tensor* args. I think there's a place for expanding it to cover tensor args and other built-in op logic (e.g. helping #92618 expand dynamic shape support, etc.), but probably in follow-up PRs. wdyt?

Reply:
Oh yeah, sorry I wasn't clear, I approved because this is OK to TODO. My point is also more general: to be super pedantically correct, we need to reason this out for each case. But it's unlikely to matter IRL